Adding to_json in prql-compiler/src/lib , and exposing to python

qharlie opened the pull request:
@max-sixty There are some python tests in prql-python/python/tests that would be nice to have run before the Release phase as a check before uploading, I could not get them to run correctly.  Pinging @rbpatt2019 also he might know how to integrate running the python tests before Release in .github/workflows/python.yaml

qharlie committed:
Adding to_json in prql-compiler/src/lib , and exposing it in prql-python/src/lib with python tests in prql-python/python/tests

qharlie committed:
Fixing to_json test

max-sixty reviewed (approved):
Looks good!

It's also fine to put rust functions in the `prql-python` crate and then over time promote them into `prql-compiler`. `to_json` seems very reasonable to have in `prql-compiler` already though.

rbpatt2019 committed:
modularise python action

Moves the logic for building with Maturin into a composite action.
This action takes one input, the OS, which controls which build is
called. This allows the whole logic to be run from a single, simple
matrix os call, followed by a build step.

rbpatt2019 committed:
Support nox for testing multiple python

rbpatt2019 committed:
Correct compile to to_sql

qharlie committed:
Testing workflow

qharlie committed:
Testing workflow

max-sixty committed:
Merge branch 'main' into main

qharlie committed:
Testing workflow

qharlie committed:
Dropping verion of pytest for older versions of python

qharlie committed:
Adding maturin step before the tests

qharlie committed:
Adding maturin depdency

qharlie committed:
Pushing maturin develop to nox

qharlie committed:
Removing support for 3.6, not pyo3 compatible

qharlie committed:
Removing support for 3.6, not pyo3 compatible

qharlie committed:
Resetting

qharlie committed:
Workflow

qharlie committed:
Workflow

qharlie committed:
trying vresion change

qharlie committed:
Testing new path to wheel

qharlie committed:
Reverting wheel location

qharlie committed:
Testing noxfile change

qharlie committed:
Testing noxfile change

rbpatt2019 committed:
Remove manylinux for windows/mac in build action for maturin

rbpatt2019 committed:
Correct path to dist.

qharlie committed:
Merge https://github.com/prql/prql

qharlie committed:
Merging prql@main

qharlie committed:
Adding serde_json to dependencies

qharlie committed:
Fixing lint

qharlie committed:
Cleaning python tests

qharlie committed:
Release workflow test

max-sixty commented on .github/actions/setup_nox/action.yml:
```diff
@@ -0,0 +1,33 @@
+      with:
+        path: .nox
+        key:
+          nox-${{inputs.cache_number }}-${{ hashFiles('prql-python/noxfile.py') }}-${{ hashFiles('prql-python/requirements.txt') }}
```
FYI sometimes there are publicly available actions that embed this logic; e.g. the rust cache one we use does this. Not sure if there is a `nox` one though

rbpatt2019 replied:
So the default setup nox action that is available doesn't do caching, and is currently unmaintained with a breaking error. So I've created one of our own to bypass this!

max-sixty commented on .github/actions/build/action.yaml:
```diff
@@ -0,0 +1,35 @@
+name: build
+description: "Use maturin to build python dists of prql-python. Conditional if checks ensure only the correct OX build is called."
```
`OS`!

max-sixty commented on .github/actions/setup_nox/action.yml:
```diff
@@ -0,0 +1,33 @@
+    required: True
+    default: 0
+
+runs:
```
Interesting — maybe `test-cargo.yml` should be an action like this, rather than a workflow

rbpatt2019 replied:
Could be! I'm a big fan of composite actions - they make things so much more dry.

max-sixty replied:
I just changed the cargo one to this design!

max-sixty commented on .github/workflows/python.yml:
```diff
@@ -23,74 +23,37 @@ concurrency:
-    # if: "startsWith(github.ref, 'refs/tags/')"
-    if: false
-    needs: [macos, windows, linux]
+    if: "startsWith(github.ref, 'refs/tags/')"
```
This should presumably be `false` until we upload the token?

rbpatt2019 replied:
Yep I'm currently altering this. The problem is that `GitHub.ref` contains whatever triggered the build - so if we don't trigger on tags, then this will never evaluate to true.

max-sixty commented on prql-compiler/src/lib.rs:
```diff
@@ -45,3 +50,19 @@ pub mod internals {
+
+        Ok(())
+    }
+}
```
Great rust @charlie-sanders !

max-sixty commented on prql-python/python/tests/test_all.py:
```diff
@@ -0,0 +1,21 @@
+    assert res is not None
+
+
+if __name__ == '__main__':
```
I haven't used `nox` much, but traditionally in pytest, this isn't required — pytest will collect and run the tests

rbpatt2019 replied:
Yep, the `if` statement is not required for nox. All it's doing is running purest manually in isolated virtualenvs of different versions.

max-sixty commented on prql-python/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+pytest==7.0.0
```
FYI I've used poetry very successfully for defining requirements for the past few years (a bit heavier though, so this is maybe better at the moment)

rbpatt2019 replied:
I thought about poetry as well! It's my go to - but the reqs list was so small that I thought it was heavier. I'm also not sure how poetry would interact with Maturin - both require build sections, and I don't think pyproject.toml allows for more than one. I'll investigate!

max-sixty commented on prql-python/src/lib.rs:
```diff
@@ -15,9 +15,21 @@ pub fn to_sql(query: &str) -> PyResult<String> {
 }
 
+#[pyfunction]
+pub fn to_json(query: &str) -> PyResult<String> {
```
Really nice!

max-sixty reviewed (commented):
Really great, thanks @charlie-sanders & @rbpatt2019 

I left some nits for information, def no actions needed
